### PR TITLE
MySQLi result mode

### LINKED ISSUE #4799: getUnbufferedRow() - memory usage
Version 4.x

When using the getUnbufferedRow() for mysql results it seems that the driver does not properly apply the appropriate logic.

According to php docs, the default result mode is: MYSQLI_STORE_RESULT 

When performing unbuffered queries you need to use MYSQLI_USE_RESULT while executing the query method.

https://www.php.net/manual/en/mysqlinfo.concepts.buffering.php

A simple hack:

in file: /system/Database/MySQLi/Connection.php

Add property to class:  public $result_mode = MYSQLI_STORE_RESULT; // default

on line 329 in the execute method change:
return $this->connID->query($this->prepQuery($sql));
to:
return $this->connID->query($this->prepQuery($sql),$this->result_mode);

Then in model when you need unbuffered results you can set result mode: $model->db->result_mode = MYSQLI_USE_RESULT;

I have a query when run in buffered mode consumes 300mb. 

Then when I use getUnbufferedRow() it still consumes 100mb.

After applying the above hack and setting $model->db->result_mode = MYSQLI_USE_RESULT; it only consumes 5mb.

Perhaps there is a better fix than the one I outline above but this seems to make huge improvements in memory usage.

### PR DESCRIPTION
**Pull request by sclubricants**
**Description**
Fixes #4799

### CODE REVIEW

**Comment by sclubricants** [comment]
First time trying to contribute to anything.. hope I did this correctly!

File: system/Database/MySQLi/Connection.php [hunk 1]
```diff
@@ -62,6 +62,19 @@ class Connection extends BaseConnection
...
+    /**
+     * MySQLi constant
+     *
+     * For unbuffered queries use MYSQLI_USE_RESULT
+	 *
+	 * Default mode for buffered queries use MYSQLI_STORE_RESULT
+     *
+     * @var int
+     */
+    public $result_mode = MYSQLI_STORE_RESULT;
```
Comment (paulbalandan) [inline system/Database/MySQLi/Connection.php:76]:
```suggestion
    /**
     * MySQLi constant
     *
     * For unbuffered queries use `MYSQLI_USE_RESULT`.
     *
     * Default mode for buffered queries uses `MYSQLI_STORE_RESULT`.
     *
     * @var int
     */
    public $resultMode = MYSQLI_STORE_RESULT;
```

File: system/Database/MySQLi/Connection.php [hunk 2]
```diff
@@ -62,6 +62,19 @@ class Connection extends BaseConnection
      */
     public $mysqli;
 
+    //--------------------------------------------------------------------
+
```
Comment (paulbalandan) [inline system/Database/MySQLi/Connection.php:66]:
```suggestion

```

File: system/Database/MySQLi/Connection.php [hunk 3]
```diff
@@ -295,7 +308,7 @@ public function execute(string $sql)
         }
 
         try {
-            return $this->connID->query($this->prepQuery($sql));
+            return $this->connID->query($this->prepQuery($sql),$this->result_mode);
```
Comment (paulbalandan) [inline system/Database/MySQLi/Connection.php:311]:
```suggestion
            return $this->connID->query($this->prepQuery($sql), $this->resultMode);
```

File: user_guide_src/source/database/results.rst [hunk 4]
```diff
@@ -153,11 +153,14 @@ parameter:
 
 This method returns a single result row without prefetching the whole
 result in memory as ``row()`` does. If your query has more than one row,
-it returns the current row and moves the internal data pointer ahead.
+it returns the current row and moves the internal data pointer ahead. For 
+use with MySQLi you must set MySQLi's result mode to MYSQLI_USE_RESULT.
```
Comment (paulbalandan) [inline user_guide_src/source/database/results.rst:157]:
```suggestion
use with MySQLi you must set MySQLi's result mode to ``MYSQLI_USE_RESULT``.
```

File: user_guide_src/source/database/results.rst [hunk 5]
```diff
@@ -153,11 +153,14 @@ parameter:
...
 
 ::
 
-    $query = $db->query("YOUR QUERY");
+    $db->result_mode = MYSQLI_USE_RESULT;
+	
+	$query = $db->query("YOUR QUERY");
```
Comment (paulbalandan) [inline user_guide_src/source/database/results.rst:163]:
```suggestion
    $db->resultMode = MYSQLI_USE_RESULT;

    $query = $db->query("YOUR QUERY");
```

**Review by paulbalandan — CHANGES_REQUESTED** [review]

**Review by paulbalandan — APPROVED** [review]
Thank you for your first contribution! 🎉

A tip for your next contributions: You do not need to merge your PR branch into your fork's `develop` branch. You can just directly open your PR here so that you will not have unnecessary merge commits.

**Review by michalsn — CHANGES_REQUESTED** [review]
I'm not against adding the ability to change the result mode, but we're intentionally using the default result mode. There are many trade-offs that come with using `MYSQLI_STORE_RESULT` and I'm pretty sure that a lot of developers are unaware of them.

For a quick reference, here is an article that explains it quite good: http://www.hackingwithphp.com/9/4/9/unbuffered-queries-for-large-data-sets-mysqli_use_result

With the current approach, we can use Query Builder as we used to and don't worry about other stuff. The database is doing a "full query" and we can process the result in PHP in an "unbuffered" manner (not the full result).

The reason behind it is that the database can handle pretty much everything and problems usually occur on the PHP side when we want to iterate on a big result set. 

I understand that this change may be desired for some use cases, but this definitely needs better documentation because I can imagine that this "tiny" change (recommendation) can generate a lot of problems for other developers.

**Comment by sclubricants** [comment]
I separated the MySQLi example from the other and made some notes on some of the associated behaviors. What felt was really missing previously is how to return to default mode.

**Review by michalsn — CHANGES_REQUESTED** [review]
Thank you @sclubricants for your changes!

Last one thing. I think we should emphasize in the user guide that changing the result mode is not required for this method to work properly.

File: user_guide_src/source/database/results.rst [hunk 6]
```diff
@@ -166,6 +166,31 @@ it returns the current row and moves the internal data pointer ahead.
         echo $row->body;
     }
 
+For use with MySQLi you must set MySQLi's result mode to 
+``MYSQLI_USE_RESULT``.
```
[the remark this answers is not in the document]
Comment (sclubricants) [reply user_guide_src/source/database/results.rst:170]: According to my tests the method does not work properly without using MYSQLI_USE_RESULT. It isn't truly an unbuffered query unless you use it. I don't quite understand why this was implemented without it. I ran a query that consumed 100+mb using the current method and then after using MYSQLI_USE_RESULT it was down to 5mb. So I'm not sure what using getUnbufferedRow() is useful for without also using MYSQLI_USE_RESULT.

[the remark this answers is not in the document]
Comment (michalsn) [reply user_guide_src/source/database/results.rst:170]: It is working correctly, but maybe not how you would expect it to work. As I mentioned earlier - even without changing the result mode this method will give developers some advantages. It's about not returning the whole result at once.

The initial implementation was meant to make it possible to iterate over huge results on the PHP side, not to optimize the query on the database side.

[the remark this answers is not in the document]
Comment (sclubricants) [reply user_guide_src/source/database/results.rst:170]: The only thing I can see is that using getUnbuggeredRow() as it is will only put one row at a time into a variable. But it should be noted that the entire result set is already in php memory.

From php docs:

MYSQLI_STORE_RESULT (default) - returns a mysqli_result object with buffered result set.

MYSQLI_USE_RESULT - returns a mysqli_result object with unbuffered result set.

As I understand this is that when query is called with the default setting the entire result set is retrieved by php from the MySQL server. This allows functions like seek, row count, etc. Php can scan through the entire result set and any given row can be fetched to a variable. When calling with Use Result, the communication channel remains open and php can request a row at a time until all rows are consumed and the channel closes. Under these circumstances php is unable to get row count and seek though the data. 

So this doesn't have anything to do with optimizing the database side. In fact its less than optimal on the database side because the database retains a read on the tables.  You said "It's about not returning the whole result at once." but it clearly does return the results to php at once even if you don't fetch them all into variables.

I'm just trying to get on the same page with you here.. I'm not sure how to document the advantages of using getUnbufferedRow() without also using MYSQLI_USE_RESULT.  I guess there is a slight advantage in that your not fetching the result as variables.

[the remark this answers is not in the document]
Comment (michalsn) [reply user_guide_src/source/database/results.rst:170]: > So this doesn't have anything to do with optimizing the database side. In fact its less than optimal on the database side because the database retains a read on the tables. 

We were talking about optimizing memory usage, so from this perspective buffer uses less memory - this is what I was referring to.

> You said "It's about not returning the whole result at once." but it clearly does return the results to php at once even if you don't fetch them all into variables.

Not exactly, it returns the whole result to the buffer and that's a big difference - in the memory usage. We may not understand each other correctly here. One thing is to return the entire result (or not) to the buffer, and the other thing is to return the buffer to PHP variables.

> I'm just trying to get on the same page with you here.. I'm not sure how to document the advantages of using getUnbufferedRow() without also using MYSQLI_USE_RESULT. I guess there is a slight advantage in that your not fetching the result as variables.

You've already done this by adding a notes section to your user guide. I think that's enough.

---

Overall, I have no problems with your arguments - you're right. What I'm trying to do here is to ensure that developers will change the result mode only if they know what they're doing. That's all.

When we use buffered mode the whole result is kept in the buffer and usually, it's not a problem. Fetching data from the buffer into PHP is the moment when problems occur.

To summarize this, when we operate on large results, we can describe methods this way:

* `$query->getResult()` - the most memory consuming and may lead to errors with exhausted allowed memory size
* `$query->getUnbufferedRow()` - safe choice, you will be able to iterate over your results without sacrificing any functionality
* `$query->getUnbufferedRow() with MYSQLI_USE_RESULT` - best choice if you're concerned about memory usage, but it comes with some trade-offs

[the remark this answers is not in the document]
Comment (sclubricants) [reply user_guide_src/source/database/results.rst:170]: Thanks, I think I'm understanding now. I had originally opened an issue about this as a bug. I was really thrown off by the method name. Looking at how its implemented and reading through the guide I kind of see how it all goes together.  I have an idea of how to reword things a bit so it doesn't interfere with the current convention.  I'm thinking of presenting the use case of writing a query out to csv. This is how I came about the problem myself. I just couldn't understand why a buffered query would consume so much memory.  This feature definitely is very useful.. but yeah you wouldn't want to use it in most circumstances.  
I'll put something together for the userguide tomorrow.

**Review by paulbalandan — CHANGES_REQUESTED** [review]
@sclubricants please rebase against `4.2` branch then force push

**Comment by sclubricants** [comment]
> @sclubricants please rebase against `4.2` branch then force push

I'm not sure if I did this correctly..

**Comment by paulbalandan** [comment]
@sclubricants your rebase went wrong. I'll try to give you the steps to clean this.
(I'm assuming here you're working locally.)

1. Checkout to `develop`: `git checkout develop`
2. Create remote for CI4 repo: `git remote add upstream https://github.com/codeigniter4/CodeIgniter4`
3. Fetch upstream: `git fetch --prune upstream`
4. Create the `4.2` branch: `git branch 4.2 upstream/4.2`
5. Create your PR branch: `git branch pr4812 4.2`
6. Checkout: `git checkout pr4812`
7. Push to your remote: `git push -u origin pr4812`
8. Cherry pick your commits:
```
git cherry-pick bf8e22627fa8d009f89162b22f60c32cc4fc04e8
git cherry-pick 0ade28e92036117b6df05ab399162f5ab7005741
git cherry-pick 5139c2472079fe460036a54161c21a3e329068a5
git cherry-pick 90b41a014b330760741df0c55d57465998d41ae7
git cherry-pick 9453c7c08f48fb094888a54663c22725b9752717
```
9. Make sure the changes are committed. Then push: `git push`
10. Open a PR directly here in the repo. NO NEED TO OPEN A PR in your fork. That will just add unnecessary commits.

**Comment by sclubricants** [comment]
I"m completely lost.

$ git push
fatal: The current branch pr4812 has no upstream branch.
To push the current branch and set the remote as upstream, use

    git push --set-upstream origin pr4812

I was using the desktop program. I have no idea about the command line. Can I just start with a clean code base and make the changes?

**Comment by paulbalandan** [comment]
Hmm.. I think you missed step 7.

Well, anyway, it will be faster to open a new PR just with your commits. Just remember to open it against `4.2` and don't make a merge commit in your fork. PR it directly here.

**Comment by kenjis** [comment]
New PR:  #4850
